Add:Sparse Transfer Learning(STL) Recipe + Tutorial

rahul-tuli opened the pull request:
* Resnet
* MobileNet

rahul-tuli committed:
Add:Sparse Transfer Learning(STL) Recipe + Tutorial
* Resnet
* MobileNet

* Apply suggestions from code review

Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>
Co-authored-by: Jeannie Finks <74554921+jeanniefinks@users.noreply.github.com>

kevinaer commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+
+---
+# Epoch and Learning-Rate variables
+num_epochs: &num_epochs 10.0
```
Should we be including this in a code block?

bfineran replied:
this is a recipe, so the YAML portion is in the front matter (enclosed by `---`)

rahul-tuli replied:
Will leave it as is for now. Will agree the codeblock makes rendered md look way nicer.

kevinaer commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+    init_lr: *init_lr
+
+# Phase 1 Sparse Transfer Learning / Recovery
+sparse_transfer_learning_modifiers:
```
Same as above regarding code block

rahul-tuli replied:
Same as above!

jeanniefinks commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+# Transfer Learning
+
+This recipe provides a framework for performing sparse transfer learning on
+SparseML's `sparsified` ResNet50, or MobileNet model.  This recipe was written for transfer learning
```
```suggestion
SparseML's `sparsified` ResNet-50, or MobileNet model.  This recipe was written for transfer learning
```

jeanniefinks commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+for your dataset and training environment.
+
+## Imagenette
+[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from Imagenet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
```
```suggestion
[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from ImageNet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
```

jeanniefinks commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+
+## Imagenette
+[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from Imagenet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
+To transfer learn Resnet50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```
```suggestion
To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```

jeanniefinks commented on integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:
```diff
@@ -0,0 +1,159 @@
+
+`ConstantPruningModifier` maintains model's sparsity levels during transfer learning process.
+We include this example [recipe](https://github.com/neuralmagic/sparseml/blob/main/integrations/pytorch/recipes/sparse-transfer-learn.md) in the SparseML [GitHub repository](https://github.com/neuralmagic/sparseml).
+To learn more about recipes and modifiers refer [here](https://github.com/neuralmagic/sparseml/blob/main/docs/source/recipes.md)
```
```suggestion
[Learn more about recipes and modifiers](https://github.com/neuralmagic/sparseml/blob/main/docs/source/recipes.md).
```

jeanniefinks commented on integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:
```diff
@@ -0,0 +1,159 @@
+Neural Magic sparse models and recipes simplify the sparsification process by enabling sparse transfer learning to create highly accurate pruned image classification models. 
+In this tutorial, you downloaded a pre-sparsified model, applied a Neural Magic recipe for sparse transfer learning, and saved the model for future use.
+
+Now, refer [here](https://github.com/neuralmagic/deepsparse/tree/main/examples/classification) for an example for benchmarking and deploying image classification models with DeepSparse.
```
```suggestion
 An example for benchmarking and deploying image classification models with DeepSparse [is also available](https://github.com/neuralmagic/deepsparse/tree/main/examples/classification).
```

rahul-tuli commented:
Thank you @jeanniefinks!

jeanniefinks reviewed (dismissed):
🚀

bfineran commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+sparse_transfer_learning_modifiers:
+  - !ConstantPruningModifier
+    start_epoch: 0.0
+    params: __ALL__
```
this will only grab the conv/FC weights and ignore stuff like batch norm

```suggestion
    params: __ALL_PRUNABLE__
```

bfineran commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+
+## Imagenette
+[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from ImageNet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
+To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```
```suggestion


To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```

bfineran commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+Some considerations:
+* For more complex datasets, increate the number of epochs, adjusting the learning rate step accordingly
+* Adding more learning rate step milestones can lead to more jumps in accuracy
+* Increasing the number of epochs before starting quantization can give a higher baseline
+* If accuracy is not recovering in the quantization phase, the total number of epochs should be increased as well
```
nit: remove if recipe does not include quantization

rahul-tuli replied:
Eye of the tiger! 🐅

bfineran commented on integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:
```diff
@@ -0,0 +1,159 @@
+Neural Magic’s ML team creates sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo. 
+Sparsifying involves removing redundant information from neural networks using algorithms such as pruning and quantization, among others. 
+This sparsification process results in many benefits for deployment environments, including faster inference and smaller file sizes. 
+Unfortunately, many have not realized the benefits due to the complicated process and number of hyperparameters involved.
```
```suggestion

Unfortunately, many have not realized the benefits due to the complicated process and number of hyperparameters involved.
```

bfineran reviewed (dismissed):
Good work @rahul-tuli! Left a few suggestions. The only procedural thing that could be changed is that the `vision.py` script can actually take SparseZoo stubs as a checkpoint input, but I think it's also nice to have the weight download as part of the tutorial...

bfineran reviewed (commented):
one more thing, let's remove the quantized and base models from the table of SparseZoo models since for the recipe to work, the user will specifically need to use a FP32 sparse model checkpoint

rahul-tuli committed:
Remove:Quantized and Base model Stubs from tutorial

markurtz commented on integrations/pytorch/recipes/sparse-transfer-learn.md:
```diff
@@ -0,0 +1,98 @@
+<!--
```
can we rename this to follow standards newly setup in the yolov5 integration? Specifically [model].transfer_learn_pruned.md and for model let's mark it as generic for classification by putting in 'classification' in place

rahul-tuli replied:
Resolved!

markurtz commented on integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:
````diff
@@ -0,0 +1,157 @@
+pip install sparseml[torchvision]
+```
+Note:  make sure to upgrade `pip` using `python -m pip install -U pip`
+## Downloading and Preparing a Pre-Sparsified Model
````
We should be able to get by without having the user run the SparseZoo code to download. Specifically the vision.py script should accept the stubs to start from and auto download from the zoo, see the YOLOv5 tutorial as an example

rahul-tuli replied:
Resolved!

markurtz commented on integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:
```diff
@@ -0,0 +1,157 @@
+## Applying a Sparse Transfer Learning Recipe
+
+- Once the model checkpoint has been downloaded, [vision.py](https://github.com/neuralmagic/sparseml/blob/main/integrations/pytorch/vision.py) script can be used to download [Imagenette](https://github.com/fastai/imagenette) and kick-start transfer learning. 
+   The transfer learning process itself is guided using recipes; a minimal example recipe could look like the following:
```
I'm not sure if we should include an example recipe in here. If they want to learn more about what the recipes look like, then we should have an external tutorial and the external recipes should contain enough information. This is going to additionally compound once we add in the quantization sparse transfer learning

rahul-tuli replied:
Removed!

rahul-tuli replied:
Okay, removed recipes from tuotrial!

markurtz reviewed (changes requested):
A few comments. Additionally for the overall structure we should follow the numbered step setup within each sub category like what was done for v3 and v5 tutorials

rahul-tuli committed:
Add: classification.transfer_learn_pruned_quantized.md
Refactor: recipe names to conform to latest integration standard

rahul-tuli committed:
Add: Quantized model stubs for resnet50 and mobilenet in classification_sparse_transfer_learning_tutorial.md

rahul-tuli committed:
Remove: SparseZoo code from classification_sparse_transfer_learning_tutorial.md
Add: command to auto-download model checkpoints from SparseZoo with custom recipes
Add: recipe stub table
Update: argument description in vision.py for clarity

rahul-tuli committed:
Merge branch 'main' into add-STL-tutorial

rahul-tuli commented:
Thank you Sensei @markurtz @bfineran ![](https://media0.giphy.com/media/DdScANHfXnsM0uoiwV/giphy.gif)
